### PR TITLE
feat: invalidate database when scraping rules change

### PR DESCRIPTION
**Pull request by billlevine**
## Proposed Changes

- New Db schema to store scraping meta data
- Include a hash of the rules used to scrape in the Db
- When opening the Db, if the hash for the rules is different, invalidate the Db so it will be re-created

## Release Notes

- New schema for the package Db will cause existing databases to be re-created.

### CODE REVIEW

**Comment by tomberek** [comment]
If we template the rules into the flake files (some of it must be in there I presume to adjust Nixpkgs evaluation, rather than just post-processing), then we can use the stock standard fingerprint of that resulting definition; invalidating automatically and preserving a 1-1 relationship with the evaluation cache.

**Comment by mkenigs** [comment]
> If we template the rules into the flake files (some of it must be in there I presume to adjust Nixpkgs evaluation, rather than just post-processing), then we can use the stock standard fingerprint of that resulting definition; invalidating automatically and preserving a 1-1 relationship with the evaluation cache.

That's contrary to the design discussed in https://github.com/flox/flox/pull/1095#issuecomment-1973266142

File: pkgdb/src/pkgdb/input.cc [hunk 1]
```diff
@@ -72,12 +73,19 @@ PkgDbInput::init()
...
       nix::logger->log(
         nix::lvlTalkative,
-        nix::fmt( "Clearing outdated database '%s'", this->dbPath.string() ) );
+        nix::fmt( "Outdated database '%s'", this->dbPath.string() ) );
+      nix::logger->log( nix::lvlTalkative,
+                        nix::fmt( "Clearing due to %s",
+                                  rulesMatch ? "table schema being outdated"
+                                             : "scraping rules mismatch" ) );
       std::filesystem::remove( this->dbPath );
       PkgDb( this->getFlake()->lockedFlake, this->dbPath.string() );
```
Comment (mkenigs) [inline pkgdb/src/pkgdb/input.cc:90]: question: does this invalidate and re-scrape immediately or on the next run? These 2 lines are making me think it happens immediately but your comments make me think it happens on the next run

File: pkgdb/tests/pkgdb.cc [hunk 2]
```diff
@@ -969,8 +970,19 @@ test_getPackages_semver0( flox::pkgdb::PkgDb & db )
...
 test_RulesTree_parse0()
 {
-  flox::pkgdb::RulesTreeNode rules(
-    rulesJSON.get<flox::pkgdb::ScrapeRulesRaw>() );
+  flox::pkgdb::ScrapeRules rules( rulesJSON );
+  return true;
+}
+
+/**
+ * @brief Ensure parsing @a flox::pkgdb::RulesTreeNode from JSON doesn't throw.
```
Comment (mkenigs) [inline pkgdb/tests/pkgdb.cc:978]: nit: copy pasted comment?

File: pkgdb/src/pkgdb/schemas.hh [hunk 3]
```diff
@@ -20,10 +20,14 @@ static const char * sql_versions = R"SQL(
 CREATE TABLE IF NOT EXISTS DbVersions (
   name     TEXT NOT NULL PRIMARY KEY
 , version  TEXT NOT NULL
+);
+
+CREATE TABLE IF NOT EXISTS DbScrapeMeta (
+  key      TEXT NOT NULL PRIMARY KEY
+, value    TEXT NOT NULL
```
Comment (mkenigs) [inline pkgdb/src/pkgdb/schemas.hh:27]: nit: wouldn't hurt to update the comment

**Review by mkenigs — CHANGES_REQUESTED** [review]

File: pkgdb/tests/migrate.bats [hunk 4]
```diff
@@ -69,6 +77,61 @@ get_version() {
...
+@test "invalidation based on rules hash" {
+  # Get the package count and hash to start, change the stored hash, 
+  # and delete the packages
+  
+  search_params="$(genParamsNixpkgsFlox)"
+  run sh -c "$PKGDB_BIN search '$search_params'"
+  assert_output --partial 'pkg1'
+  
+  # Find the db path this was scraped to
+  dbpath=$($PKGDB_BIN list | grep $TEST_HARNESS_FLAKE | awk '{ print $2 }' )
```
[the remark this answers is not in the document]
Comment (billlevine) [reply pkgdb/tests/migrate.bats:95]: Nice!

File: pkgdb/src/pkgdb/input.cc (hunk 1 above)
Comment (billlevine) [reply pkgdb/src/pkgdb/input.cc:90]: That's a good question I should resolve.  That confused me as well, but I thought I was just missing something and followed the same pattern.  It behaves as if it invalidates it for the NEXT run (confirmed by tests), but it reads as though it's immediate.  Will look further.

File: pkgdb/include/flox/pkgdb/read.hh [hunk 5]
```diff
@@ -93,13 +93,25 @@ struct SqlVersions
...
 }; /* End struct `SqlVersions' */
 
+/** @brief SQLite3 schema fields for DbScrapeMeta table. */
+struct ScrapeMeta
+{
+  /**
+   * The hash representing the Scraping rules used to generate the database.
+   * This is checked upon opening and if they differ than the rules expected in
+   * the running version of pkgdb, the datbase is invalidated so it will be
+   * recreated on the *next* invocation.
```
Comment (mkenigs) [inline pkgdb/include/flox/pkgdb/read.hh:103]: nit: this comment is outdated now, right?

**Review by mkenigs — APPROVED** [review]
Nice, looks cleaner to be re-generating on first run!

File: pkgdb/include/flox/pkgdb/read.hh (hunk 5 above)
Comment (billlevine) [reply pkgdb/include/flox/pkgdb/read.hh]: Yep, nice catch!

**Review by mkenigs — APPROVED** [review]

**Pull request merged**